sysboundary/ionosphere.cpp

```diff
@@ -851,13 +879,288 @@ namespace SBC {
+                     const Real VY 
+                     =           parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::VYCRD] 
+                     + (j + 0.5)*parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVY];
+                     const Real VZ 
+                     =           parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::VZCRD] 
+                     + (k + 0.5)*parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVZ];
+                     
+                     const Real DV3 
+                     = parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVX]
+                     * parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVY] 
+                     * parameters[vn * BlockParams::N_VELOCITY_BLOCK_PARAMS + BlockParams::DVZ];
+                     
+                     const Real normV = sqrt(VX*VX + VY*VY + VZ*VZ);
+                     const Real VdotB_norm = (B[0]*VX + B[1]*VY + B[2]*VZ)/normV;
+                     Real countAndGate = floor(VdotB_norm/cosAngle);  // gate function: 0 outside loss cone, 1 inside
+                     countAndGate = max(0.,countAndGate);
+
+                     const Real energy = 0.5 * mass * normV*normV / (1000 * physicalconstants::CHARGE); // in keV, as the limits of the bins
+                     
+                     // Find the correct energy bin number to update
+                     int binNumber = round((log(energy) - log(productionMinProtonEnergy)) / log(productionMaxProtonEnergy/productionMinProtonEnergy) * (productionNumProtonEnergies-1));
+                     binNumber = max(binNumber, 0); // anything < emin goes to the lowest channel
+                     binNumber = min(binNumber, productionNumProtonEnergies-1); // anything > emax goes to the highest channel
+                     
+                     // TODO check the bin number or -1
+                     
+                     thread_lossCone_sum.at(binNumber) += block_data[vn * SIZE_VELBLOCK + cellIndex(i,j,k)] * countAndGate * normV*normV * DV3;
+                     thread_count.at(binNumber) += countAndGate * DV3;
+                  }
+               }
+               
+               // Accumulate contributions coming from this velocity block to the 
+               // spatial cell velocity moments. If multithreading / OpenMP is used, 
+               // these updates need to be atomic:
+               # pragma omp critical
+               {
+                  for (int i=0; i<productionNumProtonEnergies; i++) {
+                     spectrum.at(i) += thread_lossCone_sum.at(i);
+                     sumWeights.at(i) += thread_count.at(i);
+                  }
+               }
+            }
+            
+            // Averaging within each bin and conversion to unit of part. cm^-2 s^-1 sr^-1 keV^-1
+            // What TODO with the sr?
```

Comment:
So I'd like expert opinions here, I didn't wrap my head around what to make of this `sr^-1` here and possibly I thus handle the solid angles/loss cone stuff incorrectly. Please enlighten me.